core/src/main/java/org/apache/iceberg/SnapshotProducer.java

```diff
@@ -195,18 +183,26 @@ protected void validate(TableMetadata currentMetadata) {
    * @param metadataToUpdate the base table metadata to apply changes to
    * @return a manifest list for the new snapshot.
    */
-  protected abstract List<ManifestFile> apply(TableMetadata metadataToUpdate);
+  protected abstract List<ManifestFile> apply(TableMetadata metadataToUpdate, Long snapshotId);
 
   @Override
   public Snapshot apply() {
     refresh();
-    Long parentSnapshotId = base.ref(targetBranch) != null ? base.ref(targetBranch).snapshotId() : null;
-    long sequenceNumber = base.nextSequenceNumber();
+    Long parentSnapshotId = null;
+    if (targetBranch != null) {
+      SnapshotRef branch = base.ref(targetBranch);
+      if (branch != null) {
+        parentSnapshotId = branch.snapshotId();
+      }
+      else {
+        parentSnapshotId = base.currentSnapshot().snapshotId();
```

Comment:
what if the `base.currentSnapshot() == null` ?

Reply:
Yes we should. I just caught this myself :) . the check is still required for the case where the table is empty, and we're producing a snapshot on a new branch. otherwise we'd hit an NPE. Fixed, thanks!

Reply:
will add a test for this